### PR TITLE
Optimize and Fix provider price fetches for sold securities and batch queries

### PR DESCRIPTION
**Pull request by wps260**
Three distinct bugs caused the price provider API to be called unnecessarily on every investment account sync.

1. Sold securities triggered a provider call on every sync forever

   import_security_prices passed end_date: Date.current for every security ever traded. Security::Price::Importer short-circuits via all_prices_exist? only when persisted_count == expected_count, where:

     expected_count = (clamped_start_date..Date.current).count

   This range increases daily, so a security closed two years ago would have all historical prices in the DB unnecessarily.  This also causes any closed securities to fetch prices daily, forever.

   Fix: separate currently-held securities (end_date: Date.current) from historical-only securities (end_date: last holding date for that security). Once a closed position's price range is complete through its last holding date, all_prices_exist? becomes permanently stable and no further provider calls occur for that security.

   "Currently held" is defined as appearing in account.current_holdings, which returns the most recent holding per security with qty != 0. On the first sync after a sell, the pre-sale holding is still the most recent, so the security correctly receives end_date: Date.current for one final sync before the new qty=0 holding is materialised.

2. Offline securities were not filtered

   account.trades.map(&:security) returned all securities regardless of the offline flag. This results in fetching of securities even if the provider cannot serve them, or if the user don't want them served for some reason (eg when there are symbol collisions that causes the wrong prices to be returned) The global MarketDataImporter correctly uses Security.online; the account-scoped importer did not.

   Fix: Security.online.where(id: all_security_ids) matches the established contract. Offline IDs still pass through the pluck step but resolve to nil in the securities hash and are skipped by the existing `next unless security` guard.

3. N+1 queries for security loading and per-security start dates

   - account.trades.map(&:security): triggered one SQL query per trade to load the security association (N+1).
   - first_required_price_date(security): issued 2 DB queries per security - one MIN(entries.date) and one EXISTS - so S securities = 2S queries.

   Fix: replace with batch queries totalling 4 regardless of security count:
   - account.current_holdings.pluck(:security_id) - current security IDs
   - account.trades.pluck(:security_id).uniq - traded security IDs
   - Security.online.where(id: ...) - load all security records at once
   - batch_first_required_price_dates: one GROUP BY security_id MIN(entries.date) over trades, one pluck for provider-holding security IDs, one GROUP BY security_id MAX(date) over holdings for historical end dates

## Summary by CodeRabbit

* **Bug Fixes**
  * Import end-dates now respect per-security state: sold positions are capped at their last holding date, while repurchased/reopened positions are treated as active through today.

* **Performance Improvements**
  * Date calculations batched to avoid redundant work and speed up price syncing.

* **Tests**
  * Added unit tests verifying end-date behavior and that offline securities are skipped during price fetches.

### CODE REVIEW

File: app/models/account/market_data_importer.rb [hunk 1]
```diff
@@ -45,34 +45,52 @@ def import_exchange_rates
...
+    last_holding_date = account.holdings
+                               .where(security_id: historical_ids)
+                               .group(:security_id)
+                               .maximum(:date)
+
+    all_security_ids.each do |security_id|
+      security = securities[security_id]
+      next unless security
+
+      end_date = current_security_ids.include?(security_id) ? Date.current : (last_holding_date[security_id] || Date.current)
```
[the remark this answers is not in the document]
Comment (wps260) [reply app/models/account/market_data_importer.rb]: fixed

File: app/models/account/market_data_importer.rb [hunk 2]
```diff
@@ -45,34 +45,52 @@ def import_exchange_rates
...
+                                             .where(security_id: security_ids)
+                                             .where.not(account_provider_id: nil)
+                                             .pluck(:security_id)
+                                             .to_set
+
+      security_ids.each_with_object({}) do |security_id, hash|
+        trade_date   = trade_start_dates[security_id]
+        holding_date = provider_holding_security_ids.include?(security_id) ? account.start_date : nil
+        hash[security_id] = [ trade_date, holding_date ].compact.min || account.start_date
+      end
```
[the remark this answers is not in the document]
Comment (wps260) [reply app/models/account/market_data_importer.rb:112]: fixed

**Comment by wps260** [comment]
@coderabbitai full review

File: app/models/account/market_data_importer.rb [hunk 3]
```diff
@@ -45,34 +45,71 @@ def import_exchange_rates
...
+
+    start_dates    = batch_first_required_price_dates(all_security_ids)
+    historical_ids = traded_security_ids - current_security_ids.to_a
+
+    # For securities no longer held, cap end_date at the last holding date so
+    # all_prices_exist? stays stable and we don't call the provider every sync.
+    last_holding_date = account.holdings
+                               .where(security_id: historical_ids)
+                               .group(:security_id)
+                               .maximum(:date)
```
[the remark this answers is not in the document]
Comment (wps260) [reply app/models/account/market_data_importer.rb:64]: Probably not worth fixing:
* No correctness risk in existing implementation, just risk of over-fetching in edge cases.
* provider_holding_security_ids  is pre-existing, and I think the update to actually might introduce an under-fetching risk.

Maintainer or SME can chime in if you think AI is more correct than me

**Review by jjmata — APPROVED** [review]

**Comment by jjmata** [comment]
Nit-picks from Sonnet 4.7:

---

## PR #1580 — Optimize and Fix provider price fetches for sold securities and batch queries

### Overview

This PR fixes three real bugs in `Account::MarketDataImporter#import_security_prices` and replaces an O(2S) query pattern with a fixed set of batch queries. The problems are well-described and the fixes are targeted.

---

### Code Quality

**Strengths:**
- The core logic is correct. Capping `end_date` to the last holding date for sold securities is exactly the right fix — it makes `all_prices_exist?` permanently stable for closed positions.
- The `reopened` detection (`trade_date > holding_date`) correctly handles the ordering dependency: `import_market_data` runs before `materialize_balances`, so stale holdings can hide a new repurchase. This edge case is subtle and the solution is clean.
- Filtering to `Security.online` now matches the contract used in the global `MarketDataImporter`.
- The `next unless security` guard already existed; offline IDs gracefully fall through.

---

### Specific Issues

**1. Inaccurate query count in method comment**

The comment in `batch_first_required_price_dates` says "3 queries total" which refers only to that method. But `import_security_prices` runs 2 more grouped queries (`last_holding_date`, `latest_trade_date`) on top of that — for a total of 7 fixed queries in the worst case. The PR description says "4 queries" which is also wrong. The comment should either be removed or updated to reflect the full count.

**2. `historical_ids` includes offline security IDs**

```ruby
historical_ids = traded_security_ids - current_security_ids.to_a
```

`traded_security_ids` contains all traded IDs including offline ones. The `last_holding_date` and `latest_trade_date` queries are run against `historical_ids`, which may include offline IDs that will be skipped by `next unless security` anyway. This is a minor inefficiency — consider:

```ruby
historical_ids = (traded_security_ids - current_security_ids.to_a) & securities.keys
```

**3. `batch_first_required_price_dates` queries more than needed**

```ruby
trade_start_dates = account.trades.group(:security_id).minimum("entries.date")
```

This fetches trade dates for *all* securities ever traded on the account, not just those in `security_ids`. Functionally correct, but slightly wasteful when called with a filtered set. A `.where(security_id: security_ids)` filter would be more precise.

**4. Tests use `create!` instead of fixtures**

Per CLAUDE.md and project conventions, tests should use Minitest + fixtures and avoid creating objects with `Family.create!`, `Account.create!`, etc. All three new tests create families and accounts inline, bypassing fixtures entirely. Existing fixtures should be reused where possible, with test-specific supplemental data created only for the edge cases.

**5. Duplicated test setup**

All three new tests repeat the same family/account creation boilerplate. A `setup` block or shared helper would reduce duplication.

**6. Hardcoded timezone in test assertion**

```ruby
end_date: Date.current.in_time_zone("America/New_York").to_date
```

This hardcoded "America/New_York" is fragile — it'll match only because `Security::Price::Importer` internally converts to that timezone before calling the provider. The assertion implicitly tests that internal conversion rather than the date logic in *this* file. If the timezone is significant, it should be explicit in the production code, not just implied by the test.

---

### Correctness Assessment

- The `end_date` logic covers: currently held, sold-and-stable, and sold-then-reopened. All three branches look correct.
- Fallback `holding_date || Date.current` when a security has trades but no holdings (pre-first-sync) is safe.
- `all_security_ids` is a `Set` (result of `Set#|`), which is fine — it's only iterated.
- `traded_security_ids - current_security_ids.to_a` requires the `.to_a` because `Array#-` doesn't accept a Set directly. This is correct.

---

### Test Coverage

The three new tests cover the three described bugs well:
- Sold security end date capping ✅
- Reopened position before rematerialization ✅
- Offline security filtering ✅

Missing: no assertion on query count reduction (bug #3). This is acceptable since query count tests are fragile, but worth noting that N+1 elimination is only validated by inspection, not by a test.

---

### Summary

The fixes are correct and the performance improvement is meaningful. The main concerns are: inaccurate query count documentation, minor inefficiencies with `historical_ids` including offline IDs, and test style that doesn't follow the project's fixture-based convention. None of these are blockers, but the test style issue is worth addressing before merge since it sets a pattern others may follow.

**Comment by wps260** [comment]
> ### Specific Issues
> **1. Inaccurate query count in method comment**
> **2. `historical_ids` includes offline security IDs**
> **3. `batch_first_required_price_dates` queries more than needed**
> **4. Tests use `create!` instead of fixtures**
> **5. Duplicated test setup**
> **6. Hardcoded timezone in test assertion**

Validated and addressed the above AI concerns

> Missing: no assertion on query count reduction (bug #3). This is acceptable since query count tests are fragile, but worth noting that N+1 elimination is only validated by inspection, not by a test.

this test is quite unnecessary.

**Review by jjmata — APPROVED** [review]
